src/vmm_unix.ml

```diff
@@ -0,0 +1,180 @@
+   ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+   OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+  ---------------------------------------------------------------------------*)
+let pp_unix_error ppf e = Fmt.string ppf (Unix.error_message e)
+
+let err_empty_line = "no command, empty command line"
+let err_file f e = R.error_msgf "%a: %a" Fpath.pp f pp_unix_error e
+
+let rec openfile fn mode perm = try Unix.openfile fn mode perm with
+  | Unix.Unix_error (Unix.EINTR, _, _) -> openfile fn mode perm
+
+let fd_for_file flag f =
+  try Ok (openfile (Fpath.to_string f) (Unix.O_CLOEXEC :: flag) 0o644)
+  with Unix.Unix_error (e, _, _) -> err_file f e
+
+let read_fd_for_file = fd_for_file Unix.[ O_RDONLY ]
+
+let write_fd_for_file = fd_for_file Unix.[ O_WRONLY ; O_APPEND ]
+
+let null = match read_fd_for_file (Fpath.v "/dev/null") with
+  | Ok fd -> fd
+  | Error _ -> invalid_arg "cannot read /dev/null"
+
+let rec create_process prog args stdout stderr =
+  try Unix.create_process prog args null stdout stderr with
+  | Unix.Unix_error (Unix.EINTR, _, _) ->
+      create_process prog args stdout stderr
+
+let rec close fd =
+  try Unix.close fd with
+  | Unix.Unix_error (Unix.EINTR, _, _) -> close fd
+
+let close_no_err fd = try close fd with _ -> ()
+
+(* own code starts here
+   (c) 2017, 2018 Hannes Mehnert, all rights reserved *)
+
+open Vmm_core
+
+let rec mkfifo name =
+  try Unix.mkfifo (Fpath.to_string name) 0o640 with
+  | Unix.Unix_error (Unix.EINTR, _, _) -> mkfifo name
+
+let image_file, fifo_file =
+  ((fun name -> Fpath.(tmpdir / (string_of_id name) + "img")),
+   (fun name -> Fpath.(tmpdir / "fifo" / (string_of_id name))))
+
+let rec fifo_exists file =
+  try Ok (Unix.((stat @@ Fpath.to_string file).st_kind = S_FIFO)) with
+  | Unix.Unix_error (Unix.ENOENT, _, _) -> Error (`Msg "noent")
+  | Unix.Unix_error (Unix.EINTR, _, _) -> fifo_exists file
+  | Unix.Unix_error (e, _, _) ->
+      R.error_msgf "file %a exists: %s" Fpath.pp file (Unix.error_message e)
+
+let uname () =
+  let cmd = Bos.Cmd.(v "uname" % "-s") in
+  lazy Bos.OS.Cmd.(run_out cmd |> out_string)
+
+let create_tap bridge =
+  Lazy.force (uname ()) >>= fun (sys, _) ->
+  match sys with
+  | x when x = "FreeBSD" ->
+    let cmd = Bos.Cmd.(v "ifconfig" % "tap" % "create") in
+    Bos.OS.Cmd.run_out cmd |> Bos.OS.Cmd.out_string >>= fun (name, _) ->
+    Bos.OS.Cmd.run Bos.Cmd.(v "ifconfig" % bridge % "addm" % name) >>= fun () ->
+    Ok name
+  | x when x = "Linux" ->
+    let prefix = "vmmtap" in
+    let rec find_n x =
+      let nam = prefix ^ string_of_int x in
+      match Bos.OS.Cmd.run Bos.Cmd.(v "ifconfig" % nam) with
+      | Error _ -> nam
+      | Ok _ -> find_n (succ x)
+    in
+    let tap = find_n 0 in
+    Bos.OS.Cmd.run Bos.Cmd.(v "ip" % "tuntap" % "add" % "mode" % "tap" % tap) >>= fun () ->
+    Bos.OS.Cmd.run Bos.Cmd.(v "brctl" % "addif" % bridge % tap) >>= fun () ->
```

Comment:
I think you can use the `iproute2` command
```ocaml
v "ip" % "link" % "set" % tap % "master" % bridge
```
instead of the call to `brctl` (which is part of `bridge-utils` that is not included in debian by default).
May have to additionally
```ocaml
v "ip" % "link" % "set" % tap % "up"
```
but not sure if that's the case here.

Not sure how to best avoid the race condition, but I guess the `vmmtap` namespacing here works for now.